### PR TITLE
Improve keyboard scrolling with editors, contentInsets, and different keyboards

### LINKED ISSUE #19214: [regression/8.0.3] Keyboard regression bugs
### Description

Exactly as documented here:

https://github.com/dotnet/maui/issues/14173#issue-1638336366 

This issue was resolved in June last year and is now back in .NET 8.

Here are the iOS issues I see with the keyboard:

1. Once the keyboard appears, you can't access the lower part of the UI. This means you can't access any fields on the bottom half of the screen while the keyboard is visible.
2. If you tap directly into one of those lower fields before the keyboard is showing, then the keyboard comes up over the top of the field, so you can't see what you are typing.
3. When the keyboard shows, in Android the screen scrolls up from the bottom to make way for the keyboard. However in iOS the whole screen moves upwards, so you also lose the top of the screen. This is happening to me when I have a grid as the main container on the page and I have an editor stuck to the bottom of the page - I end up not being able to see the editor, and I also lose the top of the screen. So to summarise - you lose access to both the top and bottom of the screen.

All the above is different to the behaviour on Android, and native iOS apps. If you can just get it to behave exactly like Android does then there won't be any problems.

### Steps to Reproduce

Open reproduction app.
Tap the "Keyboard Problem" button.
Tap one of the top fields - soft keyboard shows.
Now try to scroll down to get to one of the bottom fields on the form
Notice bottom of form is not accessible when soft keyboard is showing.

### Link to public reproduction project repository

https://github.com/williambuchanan2/MauiNavigation

### Version with bug

8.0.3

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

7.0.101

### Affected platforms

iOS

### Affected platform versions

_No response_

### Did you find any workaround?

No

### Relevant log output

_No response_

### LINKED ISSUE #20631: [iOS] Editor does not scroll when cursor goes behind keyboard
### Description

When the text content of an editor is being edited with the soft keyboard open, the editor should scroll to keep the cursor in view (i.e. when a new line is created or the text continues onto another line), however instead, the text continues behind the soft keyboard, and is unable to be scrolled into view. 

Tested on 8.0.3, 8.0.7, and 9.0.0-preview.1.9973, all display the same behaviour. 

### Steps to Reproduce

1. Load the repro project (including setting an app identifier if testing on a physical device)
2. Start the application on an iOS device or simulator
3. Edit text in the editor (bottom half of the screen) with the soft keyboard, so that it would go behind the keyboard.
4. Notice that the view does not scroll, and you are unable to scroll the text into view without closing and reopening the keyboard. 

### Link to public reproduction project repository

https://github.com/blakepulman/MauiiOSEditorNotScrolling

### Version with bug

8.0.3 GA

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Affected platform versions

iOS 17

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #22715: Controls slide up when using the keyboard in iOS
### Description

Controls slide up when using the keyboard in iOS.  The text in the first label is no longer fully visible.
The last correct version was 8.0.6.  In version 8.0.7 the controls moved down a few millimeters. From version 8.0.10 to 8.0.40 the controls moved up about half a centimeter.

### Steps to Reproduce

See the repository.

### Link to public reproduction project repository

https://github.com/geertgeerits/MauiBugKeyboard

### Version with bug

8.0.10 SR3

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

8.0.6 SR1

### Affected platforms

iOS

### Affected platform versions

iOS tested from version 15

### Did you find any workaround?

No

### Relevant log output

_No response_

### LINKED ISSUE #24496: Picker to Entry selections causes keyboard to cover low inputs
### Description

In my iOS app, we have a large amount of pickers and entry inputs. 

When a picker is selected, without tapping ‘Done’, selecting an input that is lower on the screen than the normal keyboard height can cause the keyboard to cover it. At this point users cannot see the text that is being typed, even if they try to scroll to the bottom. 

Users can only hide the keyboard by pressing return or the dismiss keyboard button and then tap another input. Keyboard displaying will work this time, letting the page scroll up until the input is visible.

_Tapping on an entry that causes the keyboard to display and the page to scroll, and then tapping on another entry/picker after that, will work. It’s only the other way around that is an issue._

Also shown in the video
Tapping high placed entries works fine until you need to scroll to the bottom to select another entry. The keyboard is still hiding the bottom rows of inputs, and you can’t scroll down far enough to see the additional inputs.
Users won’t know there are more inputs on screen unless they know to hide the keyboard every time. 

Though I’ve seen other fixes/problems reported about keyboards covering inputs, this one may be closer to issue #23901.
Unfocused events not firing when clicking on other controls seems like a theme here. 

https://github.com/user-attachments/assets/a8ed5e18-4430-4aff-8516-a9a9d57f8234

In Xamarin, our app lets us scroll and see the bottom inputs, it also auto scrolls up if the newly selected entry is below the keyboard. 

first reproduced on maui version: 

ios                        17.5.8018/8.0.100      VS 17.11.35201.85, VS 17.10.35122.118 

tested on simulator iPadOS 17.5 10th gen and on 8th gen hardware

finally retested on this version and result is the same

ios                        17.5.8018/8.0.100      SDK 8.0.400, VS 17.11.35201.85, VS 17.10.35122.118
maui-ios                   8.0.71/8.0.100         SDK 8.0.400

### Steps to Reproduce

**Picker to Entry issue**

1. Create a new Maui app.
2. Fill the main page with pickers and entry inputs
3. Select a picker on the lower half of screen
4. don't click done
5. Select an entry on the lower half of screen.

expected result: keyboard appears, and page scrolls up to reveal the entry that is focused.

actual result: keyboard appears, and page does not scroll. 

**Low Inputs Cut off Issue** 

1. Create a new Maui app.
2. Fill the main page with pickers and entry inputs
3. Select a picker on the top part of the page (such that you have to scroll down wards to see the bottom)
4. don't click done
5. try to select an entry on the lower part of the page

expected result: page scrolls to reveal the inputs at the bottom of the page.

actual result: keyboard covers all inputs below it, page does not let you keep scrolling 

### Link to public reproduction project repository

https://github.com/Axelmdez/MauiAppKeyboardTests

### Version with bug

8.0.71 SR7.1

### Is this a regression from previous behavior?

Yes, this used to work in Xamarin.Forms

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Affected platform versions

iPadOS 17.5 

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by tj-devel709**
### Description of Change

This PR addresses a few different but related issues regarding the user experience with the keyboard on iOS devices. 

1. The first issue is that currently, keyboard scrolling logic only applies if a focused entry/editor’s cursor is obstructed by the keyboard. Not only does the keyboard scroll, but the contentInsets are adjusted on the scrolling page/control allowing us to scroll to the top and bottom even with the keyboard showing. However, if we interact with something above the keyboard, we still need to adjust the contentInsets to adjust so that we can scroll and reach the bottom of the content now covered by the showing keyboard. This PR addresses this issue.

2. The second issue is when pickers are focused, the cursor has a size of zero so when we click on a picker, the screen scrolls to where only the top of the picker is showing. This PR makes it so the whole picker is now visible.

3. The third issue this PR addresses is when the keyboard is already showing and we focus a different control that invokes a different type of keyboard, we were not calculating if we need to scroll again. This issue shows up when things like a picker are focused and then an entry with a larger numeric keyboard is focused after.

4. The final issue this PR addresses is keeping the cursor above the keyboard when we add new lines to an editor. This is done by changing the contentInset on editors (whether the editor is scrolling or not) essentially moving the bottom of the editor above the keyboard. As a result, we can have a much better experience and see the cursor as we are typing.

This PR is a continuation of https://github.com/dotnet/maui/pull/21807 and https://github.com/dotnet/maui/pull/21932 and a special thanks to @albyrock87 for seeing that keeping the editor’s cursor above the keyboard was possible and for pointing out the correct way to find the rects respective to the windows which were both crucial to this PR!

https://github.com/user-attachments/assets/77fe00f8-a54f-4c1a-b63c-127ba87d040b

Small inconsistency where the cursor may stay a little higher above the keyboard than usual in an edge case: https://github.com/dotnet/maui/issues/24635.

### Issues Fixed

Fixes #19214
Fixes #24496
Fixes #20631 
Fixes #22715

### CODE REVIEW

**Comment by albyrock87** [comment]
I know this was painful to fix, so thanks a lot for all the effort spent here!

**Comment by ramonB1996** [comment]
Thank you for your work on this! Please let me know if you need somebody to test it for you :) 

I am patiently waiting for the next SR that contains this PR

**Comment by ajay-mundi** [comment]
thanks @tj-devel709 this is super helpful

**Comment by mattleibow** [comment]
@ramonB1996 you can test this PR using the artifacts from the MAUI-public build. Just download the `nuget` artifact and use that in your project.

**Review by PureWeen — DISMISSED** [review]

**Review by PureWeen — DISMISSED** [review]

**Review by mattleibow — DISMISSED** [review]

**Comment by rmarinho** [comment]
Seems IOS test failing

```
HideSoftInputOnTappedPageTest("Editor",True)
```

**Comment by crsillerjp** [comment]
Is there an ETA for this change? We are hoping on a release in the next few days or we have to decide to manually backport it into our branch.

File: src/Core/src/Platform/iOS/KeyboardAutoManagerScroll.cs [hunk 1]
```diff
@@ -750,14 +913,20 @@ static void RestorePosition()
 		}
 
 		if (ScrolledView is not null && ScrolledView.ContentInset != UIEdgeInsets.Zero)
+		{
 			UIView.Animate(AnimationDuration, 0, UIViewAnimationOptions.CurveEaseOut, () => AnimateInset(ScrolledView, UIEdgeInsets.Zero, 0), () => { });
+		}
+
```
Comment (PureWeen) [inline src/Core/src/Platform/iOS/KeyboardAutoManagerScroll.cs:919]: Does this need an else?

**Review by PureWeen — APPROVED** [review]

**Comment by Toine-db** [comment]
Why hasn't this been released yet, meanwhile there have been two releases without this fix (sr9.1 and sr9.2)

We have multiple production releases waiting for this fix.

**Comment by jfversluis** [comment]
@Toine-db as indicated by the labels, this is available in the nightlies as well as .NET 9 RC2. With .NET 9 close to being released I'm not sure if this will be backported to .NET 8 still but I'll ask around!

**Comment by Toine-db** [comment]
> @Toine-db as indicated by the labels, this is available in the nightlies as well as .NET 9 RC2. With .NET 9 close to being released I'm not sure if this will be backported to .NET 8 still but I'll ask around!

My apologies, I assumed it was backported to Net8 because it was in de Net8 nightly. 

I tested it with the nightly build, and that works fine. But I can't release production versions for our apps with nightly or beta versions of any kind.

**Comment by jfversluis** [comment]
Ah yeah! I see what you're saying sorry.

Since we're so close to .NET 9, I'm not sure if there will be another "full" .NET 8 SR soon unless we really have to. So I'm afraid this will probably land in .NET 9.

**Pull request merged**